### PR TITLE
Fix SslStream client auth with intermediate certificates on Linux

### LINKED ISSUE #37927: HttpClient with client certificate fails with OpenSslCryptographicException on Linux
I am trying to make calls from ASP.NET Core application to [Swish payment platform](https://developer.getswish.se/merchants/) 
simulator Web API. I am using `HttpClient` to make the calls and I am attaching client certificate which is required by the API.

This is the code I am using (full source code [here](https://pescn.visualstudio.com/_git/AspNetCoreSwish?path=%2Fsrc%2FSwishClient%2FControllers%2FApiController.cs&version=GBmaster)):
```
var cert = GetCertificate("Swish_Merchant_TestCertificate_1231181189.p12", "swish");

var handler = new HttpClientHandler();
handler.ClientCertificates.Add(cert);

var client = new HttpClient(handler);

const string paymentPath = "https://mss.cpc.getswish.net/swish-cpcapi/api/v1/paymentrequests/";

var response = await client.PostAsync(paymentPath, PaymentModel.GetSampleContent());

var responseContent = await response.Content.ReadAsStringAsync();
```

Everything works fine on Windows but as soon as I switch to Docker on Linux I get the following exception during the API call:

```
An unhandled exception occurred while processing the request.
OpenSslCryptographicException: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
Unknown location

SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
Interop+OpenSsl.DoSslHandshake(SafeSslHandle context, byte[] recvBuf, int recvOffset, int recvCount, out byte[] sendBuf, out int sendCount)

AuthenticationException: Authentication failed, see inner exception.
System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)

HttpRequestException: The SSL connection could not be established, see inner exception.
System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
```
```
Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL. ---> Interop+Crypto+OpenSslCryptographicException: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
   --- End of inner exception stack trace ---
   at Interop.OpenSsl.DoSslHandshake(SafeSslHandle context, Byte[] recvBuf, Int32 recvOffset, Int32 recvCount, Byte[]& sendBuf, Int32& sendCount)
   at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteContext& context, SecurityBuffer inputBuffer, SecurityBuffer outputBuffer, SslAuthenticationOptions sslAuthenticationOptions)
```
```
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception. ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception. ---> Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL. ---> Interop+Crypto+OpenSslCryptographicException: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
   --- End of inner exception stack trace ---
   at Interop.OpenSsl.DoSslHandshake(SafeSslHandle context, Byte[] recvBuf, Int32 recvOffset, Int32 recvCount, Byte[]& sendBuf, Int32& sendCount)
   at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteContext& context, SecurityBuffer inputBuffer, SecurityBuffer outputBuffer, SslAuthenticationOptions sslAuthenticationOptions)
   --- End of inner exception stack trace ---
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Security.SslState.ThrowIfExceptional()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__47_1(IAsyncResult iar)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at SwishClient.Controllers.ApiController.Call() in /src/SwishClient/Controllers/ApiController.cs:line 36
   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```
# Additional Context

The required client certificate is signed by intermediate CA. I have tried:
- To install these intermediate CA before running the app - on Windows manually, on Docker during Docker build.
- Installing these certificates programmatically from the app before making first API call

Both methods work fine on Windows, none on Linux.

Complete source code including certificates, `Dockerfile` and further instruction in `Readme` in 
https://pescn.visualstudio.com/AspNetCoreSwish/_git/AspNetCoreSwish

I can SSH into Linux docker container running my app and successfully connect to Swish API using curl
```
curl -s -S -i -k --cert ./Certificates/Swish_Merchant_TestCertificate_1231181189.p12:swish --cert-type p12 --cacert ./CA/Swish_TLS_RootCA.pem --tlsv1.1 --header "Content-Type:application/json" https://host.docker.internal:5001/api/test --data '{ "payeePaymentReference" : "0123456789", "callbackUrl" :"https://myfakehost.se/swishcallback.cfm", "payerAlias" : "4671234768","payeeAlias" : "1231181189", "amount" : "100", "currency" : "SEK","message" : "Kingston USB Flash Drive 8 GB" }'
```

### PR DESCRIPTION
**Pull request by davidsh**
On Linux, a client TLS certificate chaining to multiple intermediate certificates wasn't
sending the complete list of intermediates because the final one in the chain wasn't
trusted. This differs from Windows behavior. This was causing the application to fail
to authenticate to the payment provider in the repro.

This PR fixes SslStream so that it will send the complete list of intermediates if the
chain build operation results in a "PartialChain" instead of a fully trusted chain.

Due to the complexity of this scenario, I didn't write a standalone CI test. But I was
able to run the original repro (using a docker container) to validate the fix.

Fixes #37927

### CODE REVIEW

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs [hunk 1]
```diff
@@ -170,8 +171,16 @@ internal static bool AddExtraChainCertificates(SafeSslHandle sslContext, X509Cha
             Debug.Assert(chain != null, "X509Chain should not be null");
             Debug.Assert(chain.ChainElements.Count > 0, "chain.Build should have already been called");
 
-            // Don't count the last item (the root)
-            int stop = chain.ChainElements.Count - 1;
+            // If not trusted then include the last item (root). Otherwise, don't include it.
```
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: It's not really trust or not, it's whether it finished with a self-signed certificate.

```suggestion
            // If the last certificate is a root certificate, don't send it. PartialChain means the last cert wasn't a root.
```

**Comment by davidsh** [comment]
>Assuming you're OK with the dependency on System.Linq

System.Net.Security.dll (where SslStream lives) already references System.Linq.  I did need to add System.Linq as a reference in order to build System.Net.Http.dll due to it including the same common source.

But System.Net.Http.dll already uses System.Net.Security.dll which brings in System.Linq. So, it doesn't really change the total number of binaries being involved.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs [hunk 2]
```diff
@@ -170,8 +171,16 @@ internal static bool AddExtraChainCertificates(SafeSslHandle sslContext, X509Cha
...
+            // If the last certificate is a root certificate, don't send it. PartialChain means the last cert wasn't a root.
+            int stop;
+            if (chain.ChainStatus.Any(s => (s.Status & X509ChainStatusFlags.PartialChain) != 0))
+            {
+                stop = chain.ChainElements.Count;
+            }
+            else
+            {
+                stop = chain.ChainElements.Count - 1;
+            }
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: Can we instead do:
```C#
int stop = chain.ChainElements.Count;
foreach (X509ChainStatus s in chain.ChainStatus)
{
    if ((s.Status & X509ChainStatusFlags.PartialChain) != 0)
    {
        stop--;
        break;
    }
}
```
That should save an allocation, avoid interface and delegate calls, and avoid the need to add the LINQ dependency.

Comment (davidsh) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: Ok.  I can do that. I was under the impression that Linq was efficient at the compiler level. Is that not true?  Is it discouraged to use in product code?

Comment (stephentoub) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs]: > Is that not true?

The C# compiler does not optimize it.

> Is it discouraged to use in product code?

In our core libraries where perf can be a factor, yes.

**Pull request merged**